Implement user export API with ndjson format

mickmak-busicat opened the pull request:
@louischan-oursky this pr is focusing on the export logic first

ref dev-2055 dev-2027 dev-2031

Progress in this pr:
- Created `POST /_api/admin/users/export` with query all user instantly, result output to cloud storage
- Created `GET /_api/admin/users/export/{ID}` endpoint returning response from `Response` schema with a presigned download url

Next pr:
- handle `csv` output format and schema

Preview:

updated preview:
https://github.com/authgear/authgear-server/pull/4751#issuecomment-2365758579

louischan-oursky commented:
I changed the base branch to `feat-user-export`

louischan-oursky commented on pkg/lib/userexport/model.go:
```diff
@@ -0,0 +1,121 @@
+type Error struct {
+	Message string `json:"message,omitempty"`
+	Reason  string `json:"reason,omitempty"`
+}
```
Remove this. You should use apierrors.Error

mickmak-busicat replied:
removed

louischan-oursky commented on pkg/lib/userexport/model.go:
```diff
@@ -0,0 +1,121 @@
+	"additionalProperties": false,
+	"properties": {
+		"format": {
+			"type": "string"
```
Add `"enum": ["ndjson", "csv"]`. Add tests for the schema. You can take reference to pkg/lib/userimport/model_test.go

mickmak-busicat replied:
added enum type
will update schema when i implement csv feature

louischan-oursky commented on pkg/lib/userexport/model.go:
```diff
@@ -0,0 +1,121 @@
+	Status      redisqueue.TaskStatus `json:"status,omitempty"`
+	Request     *Request              `json:"request,omitempty"`
+	DownloadUrl string                `json:"download_url,omitempty"`
+	Error       *Error                `json:"error,omitempty"`
```
```
Error *apierrors.Error `json:"error,omitempty"`
```

mickmak-busicat replied:
updated to use apierror

louischan-oursky commented on pkg/lib/userexport/model.go:
```diff
@@ -0,0 +1,121 @@
+	Locality      string `json:"locality"`
+	Region        string `json:"region"`
+	PostalCode    string `json:"postal_code"`
+	Country       string `json:"country"`
```
Add omitempty to all fields in Address.

mickmak-busicat replied:
added omitempty

louischan-oursky commented on pkg/lib/userexport/model.go:
```diff
@@ -0,0 +1,121 @@
+type MFA struct {
+	Emails       []string   `json:"emails"`
+	PhoneNumbers []string   `json:"phone_numbers"`
+	TOTPs        []*MFATOTP `json:"totps"`
```
Add `omitempty` to all fields in MFA

mickmak-busicat replied:
added omitempty

louischan-oursky commented on pkg/lib/userexport/model.go:
```diff
@@ -0,0 +1,121 @@
+	Mfa *MFA `json:"mfa"`
+
+	BiometricCount int `json:"biometric_count"`
+	PasskeyCount   int `json:"passkey_count"`
```
For strings, add omitempty.
For maps, add omitempty.
For slices, add omitempty.
For bools, DO NOT add omitempty. (That is, you make no changes to them. They are correct now)

mickmak-busicat replied:
added omitempty

louischan-oursky commented on pkg/lib/authn/user/model.go:
```diff
@@ -317,6 +318,36 @@ func newUserModel(
+		Authenticators: authenticators,
+	}
+}
+
```
This function should live in pkg/lib/userexport

mickmak-busicat replied:
removed this function as it is not necessary

louischan-oursky commented on pkg/admin/facade/user_export.go:
```diff
@@ -0,0 +1,19 @@
+
+func (s *UserExportFacade) ExportRecords(ctx context.Context, request *userexport.Request) *userexport.Response {
+	return s.Service.ExportRecords(ctx, request)
+}
```
Remove this facade. For now, you can just inject UserExportService in the handler.

mickmak-busicat replied:
removed the facade to use worker task similar as user import now

louischan-oursky commented on pkg/lib/authn/user/queries.go:
```diff
@@ -181,3 +182,114 @@ func (p *Queries) GetMany(ids []string, role accesscontrol.Role) (users []*model
 	return
 }
+
+func (p *Queries) GetPageForExport(offset uint64) (users []*userexport.UserForExport, err error) {
```
Move the type `userexport.UserForExport` to pkg/lib/authn/user/model.go

mickmak-busicat replied:
moved to under user model

louischan-oursky commented on pkg/lib/authn/user/store.go:
```diff
@@ -285,6 +287,28 @@ func (s *Store) QueryPage(listOption ListOptions, pageArgs graphqlutil.PageArgs)
 	return users, offset, nil
 }
 
+func (s *Store) QueryForExport(offset uint64) ([]*User, error) {
```
Take limit as argument. In general, pkg/lib/authn/user is a general package that will be depended by many packages. So it should not import userexport. userexport is a package for the user export feature, so it is not a general package.

mickmak-busicat replied:
added limit argument for the query

louischan-oursky commented on pkg/lib/userexport/service.go:
```diff
@@ -22,6 +29,60 @@ func NewLogger(lf *log.Factory) Logger {
 func (s *UserExportService) ExportRecords(ctx context.Context, request *Request) *Response {
-	// TODO: Add query logic
-	return nil
+	var total_user uint64 = 0
```
No need to count the number of users. Remove this.

mickmak-busicat replied:
removed the total count

louischan-oursky commented on pkg/lib/userexport/service.go:
```diff
@@ -28,6 +29,117 @@ func NewLogger(lf *log.Factory) Logger {
+		case model.AuthenticatorTypeTOTP:
+			record.Mfa.TOTPs = append(record.Mfa.TOTPs, &MFATOTP{
+				Secret: authenticator.TOTP.Secret,
+				URI:    authenticator.TOTP.UserID,
```
This is incorrect. If you search for `GetURI`, you will see an example of how to generate a TOTP URI in pkg/lib/authenticationflow/declarative/intent_create_authenticator_totp.go

It takes an URIOptions. For issuer, you inject HTTPOrigin and use it as issuer. For accountName, you use User.EndUserAccountID.

mickmak-busicat replied:
fixed the get uri logic

mickmak-busicat commented:
@louischan-oursky fixed comment issue
also updated the endpoint to use redis task queue to do user export
the `task /get` endpoint should be working as well

if i assume this pr included everything just before creating a tmp result file to upload, see if there is anything need to be added🙏

mickmak-busicat commented:
Added upload to cloud storage logic after export and return a presigned download url in get endpoint
i think basic function is implemented

Preview:

## User export upload to storage

https://github.com/user-attachments/assets/069eedd5-1fa3-4865-bb47-7f99633425e4

## User export get endpoint with presigned url

https://github.com/user-attachments/assets/b91470cf-03aa-4677-8ff8-25999114164b


## Error handling

<img width="1195" alt="Screenshot 2024-09-22 at 3 15 59 PM" src="https://github.com/user-attachments/assets/9fe02afd-44ce-4d21-9c53-7496a39e2b72">
<img width="1204" alt="Screenshot 2024-09-22 at 3 16 06 PM" src="https://github.com/user-attachments/assets/e03c0cbe-1972-4ba6-892c-b046924d52cc">
<img width="1330" alt="Screenshot 2024-09-22 at 3 16 55 PM" src="https://github.com/user-attachments/assets/dfd740d6-ec61-4558-b579-cf218a7e51db">




i think i will handle csv format in another pr to prevent this grows so big🙏

louischan-oursky commented on pkg/admin/transport/handler_user_export_get.go:
```diff
@@ -46,7 +48,7 @@ func (h *UserExportGetHandler) handle(w http.ResponseWriter, r *http.Request) er
 	}
 
-	response, err := userexport.NewResponseFromTask(task)
+	response, err := userexport.NewResponseFromTask(task, h.CloudStorage)
```
Keep NewResponseFromTask unchanged.

Generate the presigned url here.

mickmak-busicat replied:
moved sign url to here

louischan-oursky commented on pkg/admin/transport/handler_user_export_get.go:
```diff
@@ -21,9 +22,10 @@ type UserExportGetProducer interface {
+	AppID        config.AppID
+	JSON         JSONResponseWriter
+	UserExports  UserExportGetProducer
+	CloudStorage cloudstorage.Storage
```
Introduce a new interface for dependency injection. This is our convention.

```
type UserExportGetHandlerCloudStorage interface {
  PresignGetObject(key string, expire time.Duration) (*url.URL, error)
}
```

Of course, you need to set up wire to bind it.

mickmak-busicat replied:
added new interface for this purpose

louischan-oursky reviewed (changes requested):
- The filename does not follow the spec.
- The task ID does not follow the spec.

mickmak-busicat commented on pkg/lib/userexport/service.go:
```diff
@@ -141,61 +141,52 @@ func convertDBUserToRecord(user *user.UserForExport, record *Record) {
+	writer := io.MultiWriter(io.Discard, os.Stdout)
+
+	// Bound export loop maximum count
+	const maxPageToGet = 10000
```
i thought there is a convention for giving a bound to an infinite style loop to prevent any possibility to jump into a forever loop
if no need then let me take it away👌

mickmak-busicat replied:
removed old loop bound

mickmak-busicat commented on pkg/lib/userexport/model.go:
```diff
@@ -145,6 +145,13 @@ func NewResponseFromTask(task *redisqueue.Task) (*Response, error) {
+		}
+
 		// TODO: sign a download url from filename
 		response.DownloadUrl = result.Filename
```
moved signing url when the task completed successfully

mickmak-busicat commented on pkg/lib/userexport/service.go:
```diff
@@ -141,61 +141,52 @@ func convertDBUserToRecord(user *user.UserForExport, record *Record) {
+			recordBytes := make([]byte, 0)
+			recordBytes = append(recordBytes, []byte(recordJson)...)
+			recordBytes = append(recordBytes, []byte("\n")...)
+			writer.Write(recordBytes)
```
updated

mickmak-busicat commented on pkg/lib/userexport/service.go:
```diff
@@ -34,7 +37,7 @@ func mapGet[T string | bool | map[string]interface{}](m map[string]interface{},
 }
 
-func convertDBUserToRecord(user *user.UserForExport, record *Record) {
+func (s *UserExportService) convertDBUserToRecord(user *user.UserForExport, record *Record) {
```
updated the function signature

mickmak-busicat commented on pkg/lib/userexport/service.go:
```diff
@@ -130,9 +133,20 @@ func convertDBUserToRecord(user *user.UserForExport, record *Record) {
+			totp, err := secretcode.NewTOTPFromSecret(authenticator.TOTP.Secret)
+			if err != nil {
+				s.Logger.Warningf("Failed to get TOTP URI: %s\n", err)
+				continue
```
updated to return err from constructing the TOTP url if any

mickmak-busicat commented:
@louischan-oursky 
- Updated output filename format (sorry, i fixed this earlier here: https://github.com/authgear/authgear-server/pull/4751/commits/c6e1ab798ee10459c4c9dc90000c57eb4ff7ed5c#diff-2e3cfacaf80a00d473d5e5aaaec2a780a221c98d1a0d7f9f2862fcce0ef82574R215)
- Updated task ID prefix

Preview upload:


https://github.com/user-attachments/assets/27a5efee-cba1-4bdd-bbd9-e993f3971831

louischan-oursky commented on pkg/lib/cloudstorage/azure.go:
```diff
@@ -104,8 +105,12 @@ func (s *AzureStorage) SignedURL(name string, now time.Time, duration time.Durat
+		// Inject storage account to URL when testing on IP style host, eg: 127.0.0.1
+		IPEndpointStyleInfo: azblob.IPEndpointStyleInfo{
+			AccountName: s.StorageAccount,
+		},
```
In production, I think we do not use IP style host.

Please comment this out.

mickmak-busicat replied:
commented

louischan-oursky commented on pkg/lib/userexport/service.go:
```diff
@@ -199,8 +212,58 @@ func (s *UserExportService) ExportRecords(ctx context.Context, request *Request)
+	buffer := bytes.NewBuffer(nil)
+	if _, err := io.Copy(buffer, file); err != nil {
+		return nil, err
+	}
```
Do not read the file into memory. You can set the request body to be a reader that reads from the temp file.

mickmak-busicat replied:
changed to a file reader for the body

louischan-oursky commented on pkg/lib/userexport/model.go:
```diff
@@ -143,8 +144,13 @@ func NewResponseFromTask(task *redisqueue.Task) (*Response, error) {
-		// TODO: sign a download url from filename
-		response.DownloadUrl = result.Filename
+		if result.Filename != "" && cloudStorage != nil {
+			downloadUrl, err := cloudStorage.PresignHeadObject(result.Filename, cloudstorage.PresignGetExpiresForUserExport)
```
You should call PresignGetObject.

louischan-oursky commented on pkg/admin/transport/handler_user_export_get.go:
```diff
@@ -52,11 +58,20 @@ func (h *UserExportGetHandler) handle(w http.ResponseWriter, r *http.Request) er
 	}
 
+	// Get presigned download url when the task completed successfully
+	if response.FailedAt == nil {
```
I think you need to check the status as well. When the task is still running, FailedAt is nil but it cannot be downloaded yet.

mickmak-busicat replied:
added status check

mickmak-busicat commented:
@louischan-oursky updated the fix in comment 🙏
